### ISSUE TITLE
QUIC needs API for TLS message

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#8174]
In https://http3-explained.haxx.se/en/proc-status.html I read:

> QUIC decided to use TLS 1.3 as the foundation for the crypto and security layer to avoid inventing something new and instead lean on a trustworthy and existing protocol. However, while doing this, the working group also decided that to really streamline the use of TLS in QUIC, it should only use "TLS messages" and not "TLS records" for the protocol.
> 
> This might sound like an innocuous change, but this has actually caused a significant hurdle for many QUIC stack implementors. Existing TLS libraries that support TLS 1.3 simply do not have APIs enough to expose this functionality and allow QUIC to access it. While several QUIC implementors come from larger organizations who work on their own TLS stack in parallel, this is not true for everyone.
> 
> The dominant open source heavyweight OpenSSL for example, does not have any API for this and has not expressed any desire to provide any such anytime soon (as of November 2018).

I have not heard of any such request, nor am I sure what actually would be required.

**Comment by mattcaswell** [comment]
We did have various requests for changes during the TLS1.3 development phase in order to support QUIC. However, IIRC, the QUIC draft was not entirely stable at the time and was a bit of a moving target so we backed off from trying to get full support in for the 1.1.1 release.

I'd be interested in knowing more about what exactly is required here. One option might even be to integrate QUIC directly into OpenSSL itself.

**Comment by richsalz** [comment]
Yes, there was an issue (and a PR), but then QUIC made a *big* change it how it was going to use TLS crypto, so the items were closed.  It has changed, and the current draft is at https://datatracker.ietf.org/doc/draft-ietf-quic-tls/  It's still a draft.

FWIW, I would lean against implementing QUIC in openssl, for two and a half reasons. First, it's not the group's expertise and there's not really resources given other stuff (that's 1.5 reasons :).  The second is that QUIC is going to be evolving, etc. -- that's one of the reasons why it's based on user-space use of UDP, not hacking kernel-based TCP -- and there's a real chance of OpenSSL coming out with a version that then lags behind the rest of the world, and sticks there (esp if it's an LTS release). It would be bad if OpenSSL ossified QUIC. :)

The right solution is to implement the hooks that are needed.  It is almost impossible to do that without working with a QUIC implementation that uses OpenSSL.  (I might know of one and will ask for their input.)  I think, for now, it's early to do anything about this, until a QUIC implementor creates a PR. :)

**Comment by kroeckx** [comment]
I think there is at least confusion about what QUIC really means.
My understanding is that it's both a base protocol, and HTTP/3
over it, but it's all combined or something.

And I think at least the base protocol can live inside OpenSSL,
I'm just not sure that our current SSL_write()/SSL_read() APIs
will be good enough for some of the users.

**Comment by richsalz** [comment]
Folks here might be confused :)  It is a secure transport protocol built on top of UDP.  There are application-level uses of it, such as HTTP/3 which is HTTP using QUIC as its transport. but they are separate.

I still think it's a mistake for this project to implement QUIC.

**Comment by kaduk** [comment]
I am confident that our current SSL_write()/SSL_read() APIs will not be good enough for all QUIC users.

QUIC and HTTP are easy to conflate in one's mind's eye because the chartered deliverable of the QUIC WG is only to produce a protocol that HTTP/3 runs over, and need not give consideration to other use cases.  But there is a cleanly separate secure transport protocol that provides reliable delivery over a number of concurrent distinct data streams without head-of-line blocking across streams.  (And a bunch of other stuff, too, some of which is optional and/or less done than the core bits.)

**Comment by D4v1dH03** [comment]
> Yes, there was an issue (and a PR), but then QUIC made a *big* change it how it was going to use TLS crypto, so the items were closed.  It has changed, and the current draft is at https://datatracker.ietf.org/doc/draft-ietf-quic-tls/  It's still a draft.
> 
> FWIW, I would lean against implementing QUIC in openssl, for two and a half reasons. First, it's not the group's expertise and there's not really resources given other stuff (that's 1.5 reasons :).  The second is that QUIC is going to be evolving, etc. -- that's one of the reasons why it's based on user-space use of UDP, not hacking kernel-based TCP -- and there's a real chance of OpenSSL coming out with a version that then lags behind the rest of the world, and sticks there (esp if it's an LTS release). It would be bad if OpenSSL ossified QUIC. :)
> 
> The right solution is to implement the hooks that are needed.  It is almost impossible to do that without working with a QUIC implementation that uses OpenSSL.  (I might know of one and will ask for their input.)  I think, for now, it's early to do anything about this, until a QUIC implementor creates a PR. :)
> 

QUIC is no longer a draft now. It is now RFC 9000.

**Comment by nhorman** [comment]
So, we now have a QUIC client and are working on a QUIC server

Can someone elaborate on what the request specifically is here?  @kaduk @kroeckx @mattcaswell

**Comment by tmccombs** [comment]
> we now have a QUIC client and are working on a QUIC server

There are multiple quic implementations, both client and server, which cannot use openssl due to the lack of a suitable API. 

Even once openssl does have a working implementation of QUIC for server use cases, it doesn't really solve the problem because there still may be reasons to prefer other quic implementations, including that some such implementations are more mature. 

And some applications allow you to choose between multiple TLS implementations, and since several other options (boringssl, libressl, quictls, and wolfssl)  provide an API that can be used with (or by) a separate QUIC implementation instead of bundling their own QUIC implementation, such applications would either need to swap out the entire quic stack when using openssl, or create their own compatibility layer. It looks like currently haproxy and nginx do support a compatibility layer, but has limitations (there is also the problem of lower performance with openssl 3.x, but that is a separate issue).  See https://github.com/haproxy/wiki/wiki/SSL-Libraries-Support-Status and https://nginx.org/en/docs/quic.html.

**Comment by mattcaswell** [comment]
OpenSSL 3.5 has this API now. Closing.
